backend/src/tools/api.py

```
@@ -0,0 +1,515 @@
+        _add_endpoint_to_spec(base_url, endpoint, common_query_parameters, openapi_spec)
+    return yaml.dump(openapi_spec, sort_keys=False)
+
+def openapi_from_url(url: str) -> str:
+    try:
+        response = httpx.get(url)
+        if response.status_code == 200:
+            return yaml.dump(response.json(), sort_keys=False)
+        else:
+            raise Exception(f"Failed to get OpenAPI spec from {url}")
+    except Exception as e:
+        raise Exception(f"Failed to get OpenAPI spec from {url}: {e}")
```

Ruff (0.11.9)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)